File: src/main/java/com/exactpro/th2/infraoperator/spec/strategy/linkresolver/mq/BindQueueLinkResolver.java
/*
 * Copyright 2020-2021 Exactpro (Exactpro Systems Limited)
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.exactpro.th2.infraoperator.spec.strategy.linkresolver.mq;

import com.exactpro.th2.infraoperator.model.box.schema.link.EnqueuedLink;
import com.exactpro.th2.infraoperator.model.box.schema.link.QueueDescription;
import com.exactpro.th2.infraoperator.spec.Th2CustomResource;
import com.exactpro.th2.infraoperator.spec.link.Th2Link;
import com.exactpro.th2.infraoperator.spec.link.relation.pins.PinCouplingMQ;
import com.exactpro.th2.infraoperator.spec.strategy.linkresolver.queue.QueueName;
import com.exactpro.th2.infraoperator.spec.strategy.linkresolver.queue.RoutingKeyName;
import com.exactpro.th2.infraoperator.spec.strategy.redeploy.NonTerminalException;
import com.rabbitmq.client.Channel;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.util.ArrayList;
import java.util.List;

import static com.exactpro.th2.infraoperator.util.CustomResourceUtils.extractShortCommitHash;

public class BindQueueLinkResolver {

    private static final Logger logger = LoggerFactory.getLogger(BindQueueLinkResolver.class);

    public static void resolveBoxResource(String namespace,
                                          List<Th2Link> linkResources,
                                          Th2CustomResource affectedResource) {

        String affectedResourceName = affectedResource.getMetadata().getName();
        String commitHash = extractShortCommitHash(affectedResource);

        for (var lRes : linkResources) {
            for (var pinCouple : lRes.getSpec().getBoxesRelation().getRouterMq()) {
                var queueBunch = new QueueDescription(
                        new QueueName(namespace, pinCouple.getTo()),
                        new RoutingKeyName(namespace, pinCouple.getFrom()),
                        namespace
                );

                if (affectedResourceName.equals(pinCouple.getFrom().getBoxName())
                        || affectedResourceName.equals(pinCouple.getTo().getBoxName())) {
                    bindQueues(queueBunch, commitHash);
                }
            }
        }
    }

    public static void resolveLinkResource(String namespace,
                                           Th2Link oldLinkRes,
                                           Th2Link newLinkRes) {

        String commitHash = extractShortCommitHash(newLinkRes);

        for (var pinCouple : newLinkRes.getSpec().getBoxesRelation().getRouterMq()) {
            var queueBunch = new QueueDescription(
                    new QueueName(namespace, pinCouple.getTo()),
                    new RoutingKeyName(namespace, pinCouple.getFrom()),
                    namespace
            );
            bindQueues(queueBunch, commitHash);
        }
        removeExtinctBindings(namespace, oldLinkRes, newLinkRes, commitHash);
    }

    private static List<EnqueuedLink> convert(List<PinCouplingMQ> links, String namespace, String exchange) {
        List<EnqueuedLink> mqEnqueuedLinks = new ArrayList<>();
        for (var pinCouple : links) {
            var queueBunch = new QueueDescription(
                    new QueueName(namespace, pinCouple.getTo()),
                    new RoutingKeyName(namespace, pinCouple.getFrom()),
                    exchange
            );
            mqEnqueuedLinks.add(new EnqueuedLink(pinCouple, queueBunch));
        }
        return mqEnqueuedLinks;
    }

    private static void bindQueues(QueueDescription queue, String commitHash) {

        try {
            Channel channel = RabbitMQContext.getChannel();

            String queueName = queue.getQueueName().toString();
            var currentQueue = RabbitMQContext.getQueue(queueName);
            if (currentQueue == null) {
                logger.info("Queue '{}' does not yet exist. skipping binding", queueName);
                return;
            }
            channel.queueBind(queue.getQueueName().toString(), queue.getExchange(), queue.getRoutingKey().toString());
            logger.info("Queue '{}' successfully bound to '{}' (commit-{})",
                    queueName, queue.getRoutingKey().toString(), commitHash);
        } catch (Exception e) {
            String message = "Exception while working with rabbitMq";
            logger.error(message, e);
            throw new NonTerminalException(message, e);
        }
    }

    private static void removeExtinctBindings(String namespace,
                                              Th2Link oldLinkRes,
                                              Th2Link newLinkRes,
                                              String commitHash) {
        removeExtinctBindings(namespace,
                oldLinkRes.getSpec().getBoxesRelation().getRouterMq(),
                newLinkRes.getSpec().getBoxesRelation().getRouterMq(),
                commitHash);
    }

    public static void removeExtinctBindings(String namespace,
                                             List<PinCouplingMQ> oldLinksCoupling,
                                             List<PinCouplingMQ> newLinksCoupling,
                                             String commitHash) {

        List<EnqueuedLink> oldLinks = convert(oldLinksCoupling, namespace, namespace);
        List<EnqueuedLink> newLinks = convert(newLinksCoupling, namespace, namespace);

        oldLinks.removeIf(qlb -> newLinks.stream().anyMatch(newQlb ->
                qlb.getQueueDescription().getQueueName().equals(newQlb.getQueueDescription().getQueueName())
                        && qlb.getQueueDescription().getRoutingKey()
                        .equals(newQlb.getQueueDescription().getRoutingKey())
        ));
        try {
            Channel channel = RabbitMQContext.getChannel();

            for (var extinctLink : oldLinks) {

                var fromBox = extinctLink.getFrom();
                var toBox = extinctLink.getTo();
                QueueDescription queueBunch = extinctLink.getQueueDescription();
                var queueName = queueBunch.getQueueName().toString();
                var routingKey = queueBunch.getRoutingKey().toString();

                var currentQueue = RabbitMQContext.getQueue(queueName);
                if (currentQueue == null) {
                    logger.info("Queue '{}' already removed. skipping unbinding", queueName);
                    return;
                }
                channel.queueUnbind(queueName, queueBunch.getExchange(), routingKey);

                String infoMsg = String.format(
                        "Unbind queue '%1$s' -> '%2$s'. link {%5$s.%3$s -> %5$s.%4$s} is not active (commit-%6$s)",
                        queueName, routingKey, fromBox, toBox, namespace, commitHash
                );

                logger.info(infoMsg);
            }
        } catch (Exception e) {
            String message = "Exception while working with rabbitMq";
            logger.error(message, e);
            throw new NonTerminalException(message, e);
        }
    }
}

File: src/main/kotlin/com/exactpro/th2/infraoperator/spec/strategy/linkresolver/mq/BindQueueLinkResolver.kt
/*
 * Copyright 2020-2024 Exactpro (Exactpro Systems Limited)
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.exactpro.th2.infraoperator.spec.strategy.linkresolver.mq

import com.exactpro.th2.infraoperator.model.LinkDescription
import com.exactpro.th2.infraoperator.operator.StoreHelmTh2Op.EVENT_STORAGE_BOX_ALIAS
import com.exactpro.th2.infraoperator.operator.StoreHelmTh2Op.MESSAGE_STORAGE_BOX_ALIAS
import com.exactpro.th2.infraoperator.spec.Th2CustomResource
import com.exactpro.th2.infraoperator.spec.shared.PinAttribute
import com.exactpro.th2.infraoperator.spec.shared.pin.Link
import com.exactpro.th2.infraoperator.spec.strategy.linkresolver.createEstoreQueueName
import com.exactpro.th2.infraoperator.spec.strategy.linkresolver.createEstoreRoutingKeyName
import com.exactpro.th2.infraoperator.spec.strategy.linkresolver.createMstoreQueueName
import com.exactpro.th2.infraoperator.spec.strategy.linkresolver.queue.QueueName
import com.exactpro.th2.infraoperator.spec.strategy.linkresolver.queue.RoutingKeyName
import com.exactpro.th2.infraoperator.spec.strategy.linkresolver.queue.RoutingKeyName.ROUTING_KEY_REGEXP
import com.exactpro.th2.infraoperator.spec.strategy.redeploy.NonTerminalException
import com.exactpro.th2.infraoperator.util.CustomResourceUtils
import com.exactpro.th2.infraoperator.util.CustomResourceUtils.annotationFor
import io.github.oshai.kotlinlogging.KotlinLogging

class BindQueueLinkResolver(
    private val rabbitMQContext: RabbitMQContext,
) {
    fun resolveDeclaredLinks(resource: Th2CustomResource) {
        val namespace = resource.metadata.namespace
        val resourceName = resource.metadata.name
        val commitHash = CustomResourceUtils.extractShortCommitHash(resource)
        for (subscriberPin in resource.spec.pins.mq.subscribers) {
            val queueName = QueueName(namespace, resourceName, subscriberPin.name)
            for ((box, pin) in subscriberPin.linkTo) {
                val linkDescription = LinkDescription(
                    queueName,
                    RoutingKeyName(namespace, box, pin),
                    namespace
                )
                bindQueues(linkDescription, commitHash)
            }
            removeExtinctBindings(queueName, subscriberPin.linkTo, commitHash)
        }
    }

    fun resolveHiddenLinks(resource: Th2CustomResource) {
        val namespace = resource.metadata.namespace
        val resourceName = resource.metadata.name
        val resourceLabel = annotationFor(resource)
        val commitHash = CustomResourceUtils.extractShortCommitHash(resource)
        if (resourceName.equals(EVENT_STORAGE_BOX_ALIAS) || resourceName.equals(MESSAGE_STORAGE_BOX_ALIAS)) {
            return
        }
        // create event storage link for each resource
        val estoreLinkDescription = LinkDescription(
            createEstoreQueueName(namespace),
            createEstoreRoutingKeyName(namespace, resourceName),
            namespace
        )
        bindQueues(estoreLinkDescription, commitHash)

        val currentLinks: MutableList<Link> = ArrayList()
        val queueName = createMstoreQueueName(namespace)
        // create message store link for only resources that need it
        for ((pinName, attributes) in resource.spec.pins.mq.publishers) {
            if (checkStorePinAttributes(attributes, resourceLabel, pinName)) {
                val mstoreLinkDescription = LinkDescription(
                    queueName,
                    RoutingKeyName(namespace, resourceName, pinName),
                    namespace
                )
                bindQueues(mstoreLinkDescription, commitHash)
                currentLinks.add(Link(resourceName, pinName))
            }
        }
        removeExtinctBindings(queueName, currentLinks, commitHash, resourceName)
    }

    private fun checkStorePinAttributes(attributes: Set<String>, resourceLabel: String, pinName: String?): Boolean {
        if (!attributes.contains(PinAttribute.store.name)) {
            return false
        }
        if (attributes.contains(PinAttribute.parsed.name)) {
            K_LOGGER.warn {
                "Detected a pin: $resourceLabel:$pinName with incorrect store configuration. attribute 'parsed' not allowed"
            }
            return false
        }
        if (!attributes.contains(PinAttribute.raw.name)) {
            K_LOGGER.warn {
                "Detected a pin: $resourceLabel:$pinName with incorrect store configuration. attribute 'raw' is missing"
            }
            return false
        }
        return true
    }

    private fun bindQueues(queue: LinkDescription, commitHash: String) {
        try {
            val channel = rabbitMQContext.channel
            val queueName = queue.queueName.toString()
            val currentQueue = rabbitMQContext.getQueue(queueName)
            if (currentQueue == null) {
                K_LOGGER.info {"Queue '$queueName' does not yet exist. skipping binding" }
                return
            }
            channel.queueBind(queue.queueName.toString(), queue.exchange, queue.routingKey.toString())
            K_LOGGER.info {
                "Queue '$queueName' successfully bound to '${queue.routingKey}' (commit-$commitHash)"
            }
        } catch (e: Exception) {
            val message = "Exception while working with rabbitMq"
            K_LOGGER.error(e) { message }
            throw NonTerminalException(message, e)
        }
    }

    private fun removeExtinctBindings(
        queue: QueueName,
        currentLinks: List<Link>,
        commitHash: String,
        resName: String? = null
    ) {
        val queueName = queue.toString()
        val bindingOnRabbit = rabbitMQContext.getQueueBindings(queueName)
            ?.map { it.routingKey }
            ?.filter {
                it.matches(ROUTING_KEY_REGEXP.toRegex()) &&
                    if (resName == null) true else RoutingKeyName.fromString(it).boxName == resName
            }
        val currentBindings = currentLinks.mapTo(HashSet()) {
            RoutingKeyName(queue.namespace, it.box, it.pin).toString()
        }
        try {
            val channel = rabbitMQContext.channel
            bindingOnRabbit?.forEach {
                if (!currentBindings.contains(it)) {
                    val currentQueue = rabbitMQContext.getQueue(queueName)
                    if (currentQueue == null) {
                        K_LOGGER.info { "Queue '$queueName' already removed. skipping unbinding" }
                        return
                    }
                    channel.queueUnbind(queueName, queue.namespace, it)
                    K_LOGGER.info { "Unbind queue '$it' -> '$queueName'. (commit-$commitHash)" }
                }
            }
        } catch (e: Exception) {
            val message = "Exception while removing extinct bindings"
            K_LOGGER.error(e) { message }
            throw NonTerminalException(message, e)
        }
    }

    companion object {
        private val K_LOGGER = KotlinLogging.logger { }
    }
}
